packages/go_router/lib/src/route_data.dart

```diff
@@ -205,6 +205,109 @@ abstract class ShellRouteData extends RouteData {
   );
 }
 
+/// Base class for supporting
+/// [StatefulShellRoute](https://pub.dev/documentation/go_router/latest/go_router/StatefulShellRoute-class.html)
+abstract class StatefulShellRouteData extends RouteData {
+  /// Default const constructor
+  const StatefulShellRouteData();
+
+  /// [pageBuilder] is used to build the page
+  Page<void> pageBuilder(
+    BuildContext context,
+    GoRouterState state,
+    Widget navigator,
+  ) =>
+      const NoOpPage();
+
+  /// [builder] is used to build the widget
+  Widget builder(
+    BuildContext context,
+    GoRouterState state,
+    StatefulNavigationShell navigationShell,
+  ) =>
+      throw UnimplementedError(
+        'One of `builder` or `pageBuilder` must be implemented.',
+      );
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static StatefulShellRoute $route<T extends StatefulShellRouteData>({
```

Comment:
I think you can refactor most of the duplicated logic from ShellRouteData to a ShellRouteBaseData

Reply:
Actually most logics are slightly different, `ShellRouteData` use a Widget in pagebuilder and builder and `StatefulShellRouteData` use a StatefulNavigationShell

```
Widget builder(
      BuildContext context,
      GoRouterState state,
      Widget navigator,
    ) 
```
```
Widget builder(
      BuildContext context,
      GoRouterState state,
      StatefulNavigationShell navigationShell,  //<------here
    ) 
```

Reply:
sounds good